python/lsst/daf/butler/transfers/_yaml.py

```diff
@@ -337,28 +337,23 @@ def load(self, datastore: Optional[Datastore], *,
         # FileDatasets to ingest into the datastore (in bulk):
         fileDatasets = []
         for (datasetTypeName, run), records in self.datasets.items():
-            datasetType = self.registry.getDatasetType(datasetTypeName)
             # Make a big flattened list of all data IDs and dataset_ids, while
             # remembering slices that associate them with the FileDataset
             # instances they came from.
-            dataIds: List[DataCoordinate] = []
+            datasets: List[DatasetRef] = []
             dataset_ids: List[int] = []
             slices = []
             for fileDataset in records:
-                start = len(dataIds)
-                dataIds.extend(ref.dataId for ref in fileDataset.refs)
+                start = len(datasets)
+                datasets.extend(fileDataset.refs)
                 dataset_ids.extend(ref.id for ref in fileDataset.refs)  # type: ignore
-                stop = len(dataIds)
+                stop = len(datasets)
                 slices.append(slice(start, stop))
             # Insert all of those DatasetRefs at once.
             # For now, we ignore the dataset_id we pulled from the file
             # and just insert without one to get a new autoincrement value.
             # Eventually (once we have origin in IDs) we'll preserve them.
-            resolvedRefs = self.registry.insertDatasets(
-                datasetType,
-                dataIds=dataIds,
-                run=run,
-            )
+            resolvedRefs = self.registry.importDatasets(datasets)
             # Populate our dictionary that maps int dataset_id values from the
             # export file to the new DatasetRefs
             for fileId, ref in zip(dataset_ids, resolvedRefs):
```

Comment:
Shouldn't we be able to skip all this with UUIDs in import? It's needed for integer IDs but isn't this going to slow down import for no reason in the UUID case? Where we will have imported the UUIDs that we already had? Can we at least add a short circuit where `fileId == ref.id` ?

Reply:
I think this code is still generic and can work with both integers and UUIDs. I'll see if optimization can be added if original Id is the same as returned one.

Reply:
Looking at the code I don't think we can gain much by truing to optimize `fileId == ref.id` case. We still need to build a map from ids to refs for collection association. We can avoid overriding `FileDataset.refs` but that is probably a negligible something compared to actual database transaction. I think that while we still have two types of IDs present code structure that handles generic case should be reasonable. If there is a performance hit I'd like to see some profiling numbers before trying to improve it.

Reply:
Okay. It looks like this could all get simpler if we knew we had UUIDs but I agree that it's not worth tackling on this ticket.